### PR TITLE
[WIP] Feature/cosmos

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR adds crypto and ledger api for cosmos.

## Fixes

-

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: aea/aea_builder.py [hunk 1]
```diff
@@ -1084,45 +1079,27 @@ def _verify_or_create_private_keys(aea_project_path: Path) -> None:
...
-    if fetchai_private_key_path is None:
-        _create_fetchai_private_key(
-            private_key_file=str(aea_project_path / FETCHAI_PRIVATE_KEY_FILE)
-        )
-        agent_configuration.private_key_paths.update(FETCHAI, FETCHAI_PRIVATE_KEY_FILE)
-    else:
-        try:
-            _try_validate_fet_private_key_path(
-                str(aea_project_path / fetchai_private_key_path), exit_on_error=False
+    for identifier, private_key_path in IDENTIFIER_TO_KEY_FILES.items():
```
Comment (DavidMinarsch) [inline aea/aea_builder.py:1082]: cleanup only

File: aea/cli/common.py [hunk 2]
```diff
@@ -226,33 +221,20 @@ def _verify_or_create_private_keys(ctx: Context) -> None:
...
-        _create_ethereum_private_key()
-        aea_conf.private_key_paths.update(ETHEREUM, ETHEREUM_PRIVATE_KEY_FILE)
-    else:
-        try:
-            _try_validate_ethereum_private_key_path(ethereum_private_key_path)
-        except FileNotFoundError:  # pragma: no cover
-            raise click.ClickException(
-                "File {} for private key {} not found.".format(
-                    repr(ethereum_private_key_path), ETHEREUM,
+    for identifier, private_key_path in IDENTIFIER_TO_KEY_FILES.items():
```
Comment (DavidMinarsch) [inline aea/cli/common.py:224]: cleanup only

File: aea/crypto/ethereum.py [hunk 3]
```diff
@@ -135,19 +137,26 @@ def sign_transaction(self, transaction: Any) -> Any:
...
         :param signature: the transaction signature
+        :param is_deprecated_mode: if the deprecated signing was used
         :return: the recovered address
         """
-        signable_message = encode_defunct(primitive=message)
-        addr = Account.recover_message(
-            signable_message=signable_message, signature=signature
-        )
-        return addr
+        if is_deprecated_mode:
```
Comment (DavidMinarsch) [inline aea/crypto/ethereum.py:151]: was missing the distinction..

File: aea/crypto/helpers.py [hunk 4]
```diff
@@ -93,25 +103,94 @@ def _try_validate_ethereum_private_key_path(
             raise
 
 
-def _validate_private_key_path(private_key_path: str, ledger_id: str):
+def _try_validate_cosmos_private_key_path(
```
Comment (DavidMinarsch) [inline aea/crypto/helpers.py:106]: We should try to get rid of the helpers and just expose these methods on the crypto object @marcofavorito ...load and dump already implement most of the functionality

File: aea/crypto/ledger_apis.py [hunk 5]
```diff
@@ -42,6 +47,50 @@
 LEDGER_STATUS_UNKNOWN = "UNKNOWN"
 
 
+def _instantiate_api(identifier: str, config: Dict[str, Union[str, int]]) -> LedgerApi:
```
Comment (DavidMinarsch) [inline aea/crypto/ledger_apis.py:50]: simplifies, cleans up

File: tests/test_crypto/test_cosmos_crypto.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    ), "After creation the public key must no be None"
+
+
+def test_sign_and_recover_message():
+    """Test the signing and the recovery function for the eth_crypto."""
+    account = CosmosCrypto(PRIVATE_KEY_PATH)
+    sign_bytes = account.sign_message(message=b"hello")
+    assert len(sign_bytes) > 0, "The len(signature) must not be 0"
+    recovered_addr = account.recover_message(message=b"hello", signature=sign_bytes)
+    assert recovered_addr == account.address, "Failed to recover the correct address."
```
Comment (DavidMinarsch) [inline tests/test_crypto/test_cosmos_crypto.py:57]: Not stable, as we are not able to pin down the correct verifying key.

File: aea/crypto/cosmos.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        :return: the recovered address
+        """
+        signature_b64 = base64.b64decode(signature)
+        verifying_keys = VerifyingKey.from_public_key_recovery(
+            signature_b64, message, SECP256k1, hashfunc=hashlib.sha256,
+        )
+        # TODO: pick correct key
+        public_key = verifying_keys[0].to_string("compressed").hex()
+        address = self.get_address_from_public_key(public_key)
+        return address
```
Comment (DavidMinarsch) [inline aea/crypto/cosmos.py:152]: @steevi, we are trying to recover the address from the signature and message. In ethereum this works, but here there is some indeterminacy. Can you advice?

File: tests/test_crypto/test_cosmos_crypto.py (hunk 6 above)
Comment (DavidMinarsch) [reply tests/test_crypto/test_cosmos_crypto.py:57]: @steevi

File: aea/crypto/ledger_apis.py [hunk 8]
```diff
@@ -145,6 +147,17 @@ def ethereum_api(self) -> EthereumApi:
...
+    def has_cosmos(self) -> bool:
+        """Check if it has the cosmos API."""
+        return COSMOS in self.apis.keys()
+
+    @property
+    def cosmos_api(self) -> CosmosApi:
+        """Get the Cosmos API."""
+        assert self.has_cosmos, "Cosmos API not instantiated!"
+        return cast(CosmosApi, self.apis[COSMOS])
+
```
Comment (marcofavorito) [inline aea/crypto/ledger_apis.py:160]: I think this pattern `has_x`/`x_api` should be improved. This design is not scalable in terms of the number of ledger APIs.

File: aea/crypto/wallet.py [hunk 9]
```diff
@@ -22,10 +22,11 @@
 from typing import Dict, cast
 
 from aea.crypto.base import Crypto
+from aea.crypto.cosmos import COSMOS, CosmosCrypto
 from aea.crypto.ethereum import ETHEREUM, EthereumCrypto
 from aea.crypto.fetchai import FETCHAI, FetchAICrypto
 
-SUPPORTED_CRYPTOS = [ETHEREUM, FETCHAI]
+SUPPORTED_CRYPTOS = [COSMOS, ETHEREUM, FETCHAI]
```
Comment (marcofavorito) [inline aea/crypto/wallet.py:29]: see above

File: aea/crypto/ledger_apis.py (hunk 8 above)
Comment (DavidMinarsch) [reply aea/crypto/ledger_apis.py:160]: I agree, should be changed.

**Pull request merged**